web3/_utils/method_formatters.py

```diff
@@ -486,6 +487,58 @@ def apply_list_to_array_formatter(formatter: Any) -> Callable[..., Any]:
 ABI_REQUEST_FORMATTERS = abi_request_formatters(STANDARD_NORMALIZERS, RPC_ABIS)
 
 
+def get_revert_reason(response: RPCResponse) -> str:
+    """
+    Parse revert reason from response, return None if no revert happened.
+
+    If a revert happened, but no message has been given, return an empty string.
+
+    Reverts contain a `data` attribute with the following layout:
+        "Reverted "
+        Function selector for Error(string): 08c379a (4 bytes)
+        Data offset: 32 (32 bytes)
+        String length (32 bytes)
+        Reason strong (padded, use string length from above to get meaningful part)
+
+    See also https://solidity.readthedocs.io/en/v0.6.3/control-structures.html#revert
+    """
+    if not isinstance(response['error'], dict):
+        return None
+
+    # Parity/OpenEthereum case:
+    data = response['error'].get('data', '')
+    if data.startswith('Reverted '):
+        # "Reverted", function selector and offset are always the same for revert errors
+        prefix = 'Reverted 0x08c379a00000000000000000000000000000000000000000000000000000000000000020'  # noqa: 501
+        if not data.startswith(prefix):
+            return ''
+
+        reason_length = int(data[len(prefix):len(prefix) + 64], 16)
+        reason = data[len(prefix) + 64:len(prefix) + 64 + reason_length * 2]
```

Comment:
This all feels fragile.  That might be "ok" but it still worries me.  Any thoughts on just doing the simple check that the data contains `"Reverted "` and then skipping this parsing logic and embedding the whole response message in the revert exception?

Reply:
I might be misunderstanding, but I'd definitely like to include the parsed revert message and not put that on the end user's shoulders. Are you suggesting returning both the reason and the full error dict, just the error dict, or something else?

Reply:
Current thinking in support of :shipit: : 
- in related [issues filed](https://github.com/ethereum/web3.py/search?q=revert&type=issues), users are generally looking for the plaintext reason.
- handling reverts is broken today, so even brittle parsing logic is an improvement. 
- I don't expect clients to change this formatting on a whim. 
- we can iterate on it.

Reply:
I'm proposing that we return the string which we are trying to parse in full.  basically this whole context block would become:

```python
if data.startswith("Reverted "):
    reason = data
```

I don't feel strongly about this.  It just seems like it would be roughly the same usefulness to the user and would avoid the part that looks brittle.

Reply:
Makes good sense, but it feels like signing up for an endless trickle of "what is this? how do I parse this?" questions. I'd prefer to sign up for patching the logic on the hopefully very rare occasions required and deliver the superior UX. Will flag as this something "to harden" though.